File: brain.py
"""
Swimming Brain — RAG Engine
ChromaDB local embeddings for search + Gemini free tier for AI chat.
"""

import os
import json
import uuid
import time
import re
from pathlib import Path

import chromadb
from PyPDF2 import PdfReader
import google.generativeai as genai

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------
CHUNK_SIZE = 2048          # characters per chunk
CHUNK_OVERLAP = 200        # character overlap between chunks
COLLECTION_NAME = "swimming_sources"
TOP_K = 5
DATA_DIR = "data"
SOURCES_DIR = os.path.join(DATA_DIR, "sources")
CHROMA_DIR = os.path.join(DATA_DIR, "chroma")
CHAT_HISTORY_FILE = os.path.join(DATA_DIR, "chat_history.json")
SOURCES_INDEX_FILE = os.path.join(DATA_DIR, "sources_index.json")

# ---------------------------------------------------------------------------
# Module-level state
# ---------------------------------------------------------------------------
_chroma_client = None
_collection = None
_model = None


# ===================================================================
# INITIALIZATION
# ===================================================================

def init_brain(api_key: str = None) -> None:
    """Initialize ChromaDB (local embeddings) + Gemini (free tier AI chat)."""
    global _chroma_client, _collection, _model

    os.makedirs(SOURCES_DIR, exist_ok=True)
    os.makedirs(CHROMA_DIR, exist_ok=True)

    # ChromaDB uses built-in all-MiniLM-L6-v2 embeddings locally — no API
    _chroma_client = chromadb.PersistentClient(path=CHROMA_DIR)
    _collection = _chroma_client.get_or_create_collection(
        name=COLLECTION_NAME,
        metadata={"hnsw:space": "cosine"}
    )

    # Gemini free tier for AI chat generation
    if api_key:
        genai.configure(api_key=api_key)
        _model = genai.GenerativeModel("gemini-2.0-flash")
        print(f"  Brain initialized — {_collection.count()} chunks, Gemini AI enabled")
    else:
        print(f"  Brain initialized — {_collection.count()} chunks, no AI (search only)")


# ===================================================================
# SOURCE INDEX
# ===================================================================

def _load_sources_index() -> dict:
    if os.path.exists(SOURCES_INDEX_FILE):
        with open(SOURCES_INDEX_FILE, "r", encoding="utf-8") as f:
            return json.load(f)
    return {}


def _save_sources_index(index: dict) -> None:
    with open(SOURCES_INDEX_FILE, "w", encoding="utf-8") as f:
        json.dump(index, f, indent=2)


def add_source_to_index(source_id: str, metadata: dict) -> None:
    index = _load_sources_index()
    index[source_id] = metadata
    _save_sources_index(index)


def remove_source_from_index(source_id: str) -> None:
    index = _load_sources_index()
    index.pop(source_id, None)
    _save_sources_index(index)


def get_all_sources() -> list:
    index = _load_sources_index()
    sources = list(index.values())
    sources.sort(key=lambda s: s.get("uploaded_at", 0), reverse=True)
    return sources


def get_source(source_id: str) -> dict | None:
    index = _load_sources_index()
    return index.get(source_id)


# ===================================================================
# TEXT EXTRACTION
# ===================================================================

def extract_text_from_pdf(file_path: str) -> str:
    try:
        reader = PdfReader(file_path)
        pages = []
        for page in reader.pages:
            text = page.extract_text()
            if text:
                pages.append(text)
        return "\n\n".join(pages)
    except Exception as e:
        print(f"  PDF extraction error: {e}")
        return ""


def extract_text_from_txt(file_path: str) -> str:
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            return f.read()
    except UnicodeDecodeError:
        with open(file_path, "r", encoding="latin-1") as f:
            return f.read()


# ===================================================================
# CHUNKING
# ===================================================================

def chunk_text(text: str, chunk_size: int = CHUNK_SIZE, overlap: int = CHUNK_OVERLAP) -> list:
    text = text.strip()
    if not text:
        return []

    if len(text) <= chunk_size:
        return [text]

    chunks = []
    start = 0

    while start < len(text):
        end = start + chunk_size

        if end >= len(text):
            chunks.append(text[start:].strip())
            break

        slice_text = text[start:end]
        para_break = slice_text.rfind("\n\n")
        if para_break > chunk_size * 0.3:
            end = start + para_break + 2
        else:
            for sep in [". ", "! ", "? ", ".\n", "\n"]:
                sent_break = slice_text.rfind(sep)
                if sent_break > chunk_size * 0.3:
                    end = start + sent_break + len(sep)
                    break

        chunk = text[start:end].strip()
        if chunk:
            chunks.append(chunk)

        start = end - overlap
        if start <= (end - chunk_size):
            start = end

    return chunks


# ===================================================================
# CHROMADB STORAGE & RETRIEVAL (embeddings handled automatically)
# ===================================================================

def store_chunks(source_id: str, chunks: list, metadata: dict) -> int:
    """Store chunks in ChromaDB. Embeddings generated automatically by ChromaDB."""
    ids = [f"{source_id}_chunk_{i}" for i in range(len(chunks))]
    metadatas = [
        {
            "source_id": source_id,
            "source_title": metadata.get("title", "Unknown"),
            "source_type": metadata.get("type", "unknown"),
            "chunk_index": i,
        }
        for i in range(len(chunks))
    ]

    _collection.add(
        ids=ids,
        documents=chunks,
        metadatas=metadatas,
    )

    return len(chunks)


def delete_source_chunks(source_id: str) -> int:
    try:
        results = _collection.get(where={"source_id": source_id})
        if results["ids"]:
            _collection.delete(ids=results["ids"])
            return len(results["ids"])
    except Exception as e:
        print(f"  ChromaDB delete error: {e}")
    return 0


def retrieve_context(query: str, top_k: int = TOP_K) -> list:
    """Retrieve relevant chunks using ChromaDB's built-in semantic search."""
    if _collection.count() == 0:
        return []

    results = _collection.query(
        query_texts=[query],
        n_results=min(top_k, _collection.count()),
        include=["documents", "metadatas", "distances"]
    )

    context = []
    for i in range(len(results["ids"][0])):
        context.append({
            "text": results["documents"][0][i],
            "source_id": results["metadatas"][0][i]["source_id"],
            "source_title": results["metadatas"][0][i]["source_title"],
            "distance": results["distances"][0][i],
            "chunk_index": results["metadatas"][0][i]["chunk_index"],
        })

    return context


# ===================================================================
# CHAT HISTORY
# ===================================================================

def load_chat_history() -> list:
    if os.path.exists(CHAT_HISTORY_FILE):
        with open(CHAT_HISTORY_FILE, "r", encoding="utf-8") as f:
            return json.load(f)
    return []


def save_chat_message(role: str, text: str, citations: list = None) -> None:
    history = load_chat_history()
    history.append({
        "role": role,
        "text": text,
        "time": time.time(),
        "citations": citations or [],
    })
    with open(CHAT_HISTORY_FILE, "w", encoding="utf-8") as f:
        json.dump(history, f, indent=2)


def clear_chat_history() -> None:
    if os.path.exists(CHAT_HISTORY_FILE):
        os.remove(CHAT_HISTORY_FILE)


# ===================================================================
# RESPONSE GENERATION
# ===================================================================

SYSTEM_PROMPT = """You are Swimming Brain, an expert AI swimming coach.
You help swimmers improve technique, build training plans, and understand swimming science.

RULES:
- Answer using the PROVIDED SOURCE CONTEXT when available
- Cite sources with [Source: title] format after relevant statements
- If no sources are relevant, answer from general swimming knowledge and say so
- Be specific about body position, timing, catch mechanics, and common errors
- For workouts, include warm-up, main set, and cool-down with intervals
- Keep responses practical, actionable, and encouraging
- Use proper swimming terminology (catch, pull, recovery, streamline, EVF, DPS, etc.)
"""


def _build_citations(context_chunks: list) -> list:
    seen = set()
    citations = []
    for chunk in context_chunks:
        if chunk["source_id"] not in seen:
            seen.add(chunk["source_id"])
            citations.append({
                "source_id": chunk["source_id"],
                "source_title": chunk["source_title"],
            })
    return citations


def generate_response(query: str, context_chunks: list, chat_history: list = None) -> dict:
    """Generate response — uses Gemini if available, otherwise returns raw passages."""

    if not context_chunks:
        no_source_msg = "I don't have any sources about that yet. Upload some swimming PDFs, coach notes, or articles and I'll be able to help!"
        # If Gemini is available, let it answer from general knowledge
        if _model:
            try:
                prompt = f"User question: {query}\n\nNo uploaded sources are available. Answer from your general swimming knowledge."
                response = _model.generate_content(
                    contents=prompt,
                    generation_config=genai.GenerationConfig(temperature=0.7, max_output_tokens=2048),
                    system_instruction=SYSTEM_PROMPT,
                )
                return {"response": response.text, "citations": [], "chunks_used": 0}
            except Exception:
                pass
        return {"response": no_source_msg, "citations": [], "chunks_used": 0}

    # Filter low-relevance chunks
    relevant = [c for c in context_chunks if c["distance"] < 1.5]
    if not relevant:
        relevant = context_chunks[:2]  # fallback to top 2

    citations = _build_citations(relevant)

    # --- Gemini AI path (free tier) ---
    if _model:
        try:
            # Build context block
            context_text = "\n\n".join(
                f"[Source: {c['source_title']}]\n{c['text']}" for c in relevant
            )

            # Include recent chat history
            history_text = ""
            if chat_history:
                recent = chat_history[-6:]
                history_text = "\n".join(
                    f"{'User' if m['role'] == 'user' else 'Assistant'}: {m['text'][:500]}"
                    for m in recent
                )

            prompt = f"SOURCES:\n{context_text}\n\n"
            if history_text:
                prompt += f"RECENT CONVERSATION:\n{history_text}\n\n"
            prompt += f"User question: {query}"

            response = _model.generate_content(
                contents=prompt,
                generation_config=genai.GenerationConfig(temperature=0.7, max_output_tokens=2048),
                system_instruction=SYSTEM_PROMPT,
            )
            return {"response": response.text, "citations": citations, "chunks_used": len(relevant)}
        except Exception as e:
            print(f"  Gemini error, falling back to raw passages: {e}")

    # --- Fallback: raw passages (no AI) ---
    parts = [f"Here's what I found in your sources about **\"{query}\"**:\n"]
    for chunk in relevant:
        text = chunk["text"].strip()
        if len(text) > 800:
            text = text[:800].rsplit(" ", 1)[0] + "..."
        parts.append(f"**From: {chunk['source_title']}**")
        parts.append(text)
        parts.append("")

    return {"response": "\n".join(parts), "citations": citations, "chunks_used": len(relevant)}


# ===================================================================
# TOP-LEVEL ORCHESTRATION
# ===================================================================

def process_upload(file_path: str, filename: str, file_type: str, title: str = None) -> dict:
    source_id = uuid.uuid4().hex[:12]
    display_title = title or Path(filename).stem.replace("_", " ").replace("-", " ").title()

    if file_type == "pdf":
        text = extract_text_from_pdf(file_path)
    elif file_type == "txt":
        text = extract_text_from_txt(file_path)
    elif file_type == "image":
        metadata = {
            "id": source_id,
            "title": display_title,
            "type": "image",
            "filename": os.path.basename(file_path),
            "uploaded_at": time.time(),
            "chunk_count": 0,
        }
        add_source_to_index(source_id, metadata)
        return {"ok": True, "source": metadata}
    else:
        return {"error": f"Unsupported file type: {file_type}"}

    if not text.strip():
        return {"error": "Could not extract text. The file may be image-based."}

    chunks = chunk_text(text)
    if not chunks:
        return {"error": "No content to process."}

    metadata = {
        "id": source_id,
        "title": display_title,
        "type": file_type,
        "filename": os.path.basename(file_path),
        "uploaded_at": time.time(),
        "chunk_count": len(chunks),
    }
    store_chunks(source_id, chunks, metadata)
    add_source_to_index(source_id, metadata)

    print(f"  Processed: {display_title} — {len(chunks)} chunks stored")
    return {"ok": True, "source": metadata}


def process_note(title: str, content: str) -> dict:
    source_id = uuid.uuid4().hex[:12]
    safe_name = re.sub(r"[^a-zA-Z0-9_-]", "_", title.lower())[:50]
    filename = f"{source_id}_{safe_name}.txt"
    file_path = os.path.join(SOURCES_DIR, filename)

    full_text = f"{title}\n\n{content}"
    with open(file_path, "w", encoding="utf-8") as f:
        f.write(full_text)

    chunks = chunk_text(full_text)
    if not chunks:
        return {"error": "Note content is empty."}

    metadata = {
        "id": source_id,
        "title": title,
        "type": "note",
        "filename": filename,
        "uploaded_at": time.time(),
        "chunk_count": len(chunks),
    }
    store_chunks(source_id, chunks, metadata)
    add_source_to_index(source_id, metadata)

    print(f"  Note saved: {title} — {len(chunks)} chunks stored")
    return {"ok": True, "source": metadata}


def delete_source(source_id: str) -> dict:
    source = get_source(source_id)
    if not source:
        return {"error": "Source not found"}

    file_path = os.path.join(SOURCES_DIR, source.get("filename", ""))
    if os.path.exists(file_path):
        os.remove(file_path)

    deleted = delete_source_chunks(source_id)
    remove_source_from_index(source_id)

    print(f"  Deleted: {source.get('title')} — {deleted} chunks removed")
    return {"ok": True, "deleted": source_id, "chunks_removed": deleted}


def ask(query: str) -> dict:
    """Search sources, generate AI response (if Gemini available), save to history."""
    history = load_chat_history()
    context = retrieve_context(query)
    result = generate_response(query, context, history)

    save_chat_message("user", query)
    save_chat_message("assistant", result["response"], result["citations"])

    return result

File: js/brain.js
/* ===================================================================
   Swimming Brain — Brain Engine (client-side)
   Stores sources in localStorage, searches with keyword scoring,
   generates answers with Gemini free tier API.
   =================================================================== */

const Brain = (() => {
  const STORAGE_KEY = "sb-sources";
  const HISTORY_KEY = "sb-history";
  const APIKEY_KEY = "sb-gemini-key";
  const CHUNK_SIZE = 1500;
  const CHUNK_OVERLAP = 150;

  const SYSTEM_PROMPT = `You are Swimming Brain, an expert AI swimming coach.
You help swimmers improve technique, build training plans, and understand swimming science.

RULES:
- Answer using the PROVIDED SOURCE CONTEXT when available
- Cite sources with [Source: title] format after relevant statements
- If no sources are relevant, answer from general swimming knowledge and say so
- Be specific about body position, timing, catch mechanics, and common errors
- For workouts, include warm-up, main set, and cool-down with intervals
- Keep responses practical, actionable, and encouraging
- Use proper swimming terminology (catch, pull, recovery, streamline, EVF, DPS, etc.)`;

  // ----- API Key -----
  function getApiKey() {
    return localStorage.getItem(APIKEY_KEY) || "";
  }

  function setApiKey(key) {
    localStorage.setItem(APIKEY_KEY, key.trim());
  }

  function hasApiKey() {
    return !!getApiKey();
  }

  // ----- Source Storage (localStorage) -----
  function _loadSources() {
    try { return JSON.parse(localStorage.getItem(STORAGE_KEY)) || []; }
    catch { return []; }
  }

  function _saveSources(sources) {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(sources));
  }

  function getAllSources() {
    return _loadSources().sort((a, b) => b.uploadedAt - a.uploadedAt);
  }

  function addSource(source) {
    const sources = _loadSources();
    sources.push(source);
    _saveSources(sources);
  }

  function deleteSource(id) {
    const sources = _loadSources().filter(s => s.id !== id);
    _saveSources(sources);
  }

  // ----- Chat History -----
  function getHistory() {
    try { return JSON.parse(localStorage.getItem(HISTORY_KEY)) || []; }
    catch { return []; }
  }

  function saveMessage(role, text, citations) {
    const history = getHistory();
    history.push({ role, text, time: Date.now(), citations: citations || [] });
    localStorage.setItem(HISTORY_KEY, JSON.stringify(history));
  }

  function clearHistory() {
    localStorage.removeItem(HISTORY_KEY);
  }

  // ----- Text Chunking -----
  function chunkText(text) {
    text = text.trim();
    if (!text) return [];
    if (text.length <= CHUNK_SIZE) return [text];

    const chunks = [];
    let start = 0;

    while (start < text.length) {
      let end = start + CHUNK_SIZE;
      if (end >= text.length) {
        chunks.push(text.slice(start).trim());
        break;
      }

      const slice = text.slice(start, end);
      const paraBreak = slice.lastIndexOf("\n\n");
      if (paraBreak > CHUNK_SIZE * 0.3) {
        end = start + paraBreak + 2;
      } else {
        for (const sep of [". ", "! ", "? ", ".\n", "\n"]) {
          const sentBreak = slice.lastIndexOf(sep);
          if (sentBreak > CHUNK_SIZE * 0.3) {
            end = start + sentBreak + sep.length;
            break;
          }
        }
      }

      const chunk = text.slice(start, end).trim();
      if (chunk) chunks.push(chunk);
      start = end - CHUNK_OVERLAP;
      if (start <= end - CHUNK_SIZE) start = end;
    }

    return chunks;
  }

  // ----- PDF Extraction (pdf.js) -----
  async function extractPdfText(file) {
    const arrayBuffer = await file.arrayBuffer();
    pdfjsLib.GlobalWorkerOptions.workerSrc =
      "https://cdnjs.cloudflare.com/ajax/libs/pdf.js/3.11.174/pdf.worker.min.js";
    const pdf = await pdfjsLib.getDocument({ data: arrayBuffer }).promise;
    const pages = [];
    for (let i = 1; i <= pdf.numPages; i++) {
      const page = await pdf.getPage(i);
      const content = await page.getTextContent();
      pages.push(content.items.map(item => item.str).join(" "));
    }
    return pages.join("\n\n");
  }

  // ----- Process Upload -----
  async function processFile(file, title) {
    const id = crypto.randomUUID().slice(0, 12);
    const ext = file.name.split(".").pop().toLowerCase();
    let text = "";

    if (ext === "pdf") {
      text = await extractPdfText(file);
    } else {
      text = await file.text();
    }

    if (!text.trim()) {
      throw new Error("Could not extract text from this file.");
    }

    const chunks = chunkText(text);
    const source = {
      id,
      title: title || file.name.replace(/\.[^.]+$/, "").replace(/[-_]/g, " "),
      type: ext === "pdf" ? "pdf" : "txt",
      chunks,
      chunkCount: chunks.length,
      uploadedAt: Date.now(),
    };

    addSource(source);
    return source;
  }

  function processNote(title, content) {
    const id = crypto.randomUUID().slice(0, 12);
    const fullText = `${title}\n\n${content}`;
    const chunks = chunkText(fullText);

    const source = {
      id,
      title,
      type: "note",
      chunks,
      chunkCount: chunks.length,
      uploadedAt: Date.now(),
    };

    addSource(source);
    return source;
  }

  // ----- Search (keyword scoring) -----
  function searchSources(query, topK = 5) {
    const sources = _loadSources();
    if (!sources.length) return [];

    const queryWords = query.toLowerCase().split(/\s+/).filter(w => w.length > 2);
    const results = [];

    for (const source of sources) {
      for (let i = 0; i < source.chunks.length; i++) {
        const chunk = source.chunks[i];
        const chunkLower = chunk.toLowerCase();
        let score = 0;

        for (const word of queryWords) {
          const matches = (chunkLower.match(new RegExp(word, "g")) || []).length;
          score += matches;
        }

        // Bonus for exact phrase match
        if (chunkLower.includes(query.toLowerCase())) {
          score += 10;
        }

        if (score > 0) {
          results.push({
            text: chunk,
            sourceId: source.id,
            sourceTitle: source.title,
            score,
            chunkIndex: i,
          });
        }
      }
    }

    results.sort((a, b) => b.score - a.score);
    return results.slice(0, topK);
  }

  // ----- Gemini API Call -----
  async function callGemini(prompt) {
    const key = getApiKey();
    if (!key) throw new Error("No API key");

    const res = await fetch(
      `https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash:generateContent?key=${key}`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          system_instruction: { parts: [{ text: SYSTEM_PROMPT }] },
          contents: [{ parts: [{ text: prompt }] }],
          generationConfig: { temperature: 0.7, maxOutputTokens: 2048 },
        }),
      }
    );

    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.error?.message || `Gemini API error ${res.status}`);
    }

    const data = await res.json();
    return data.candidates?.[0]?.content?.parts?.[0]?.text || "No response generated.";
  }

  // ----- Main Ask Function -----
  async function ask(query) {
    const context = searchSources(query);
    const history = getHistory();

    // Build citations list
    const seen = new Set();
    const citations = [];
    for (const c of context) {
      if (!seen.has(c.sourceId)) {
        seen.add(c.sourceId);
        citations.push({ source_id: c.sourceId, source_title: c.sourceTitle });
      }
    }

    // Try Gemini AI
    if (hasApiKey()) {
      try {
        let prompt = "";

        if (context.length) {
          const contextText = context.map(c =>
            `[Source: ${c.sourceTitle}]\n${c.text}`
          ).join("\n\n");
          prompt += `SOURCES:\n${contextText}\n\n`;
        } else {
          prompt += "No uploaded sources match this question. Answer from general swimming knowledge.\n\n";
        }

        // Recent history
        const recent = history.slice(-6);
        if (recent.length) {
          prompt += "RECENT CONVERSATION:\n";
          prompt += recent.map(m =>
            `${m.role === "user" ? "User" : "Assistant"}: ${m.text.slice(0, 500)}`
          ).join("\n") + "\n\n";
        }

        prompt += `User question: ${query}`;

        const response = await callGemini(prompt);

        saveMessage("user", query);
        saveMessage("assistant", response, citations);

        return { response, citations, chunksUsed: context.length };
      } catch (e) {
        console.error("Gemini error:", e);
        // Fall through to offline mode
      }
    }

    // Offline fallback — return raw passages
    saveMessage("user", query);

    if (!context.length) {
      const msg = hasApiKey()
        ? "Gemini API error. Check your API key in settings."
        : "Add a Gemini API key (click the key icon) for AI answers, or upload sources to search.";
      saveMessage("assistant", msg, []);
      return { response: msg, citations: [], chunksUsed: 0 };
    }

    const parts = [`Here's what I found in your sources:\n`];
    for (const c of context) {
      let text = c.text.trim();
      if (text.length > 600) text = text.slice(0, 600).replace(/\s\S*$/, "") + "...";
      parts.push(`**From: ${c.sourceTitle}**\n${text}\n`);
    }

    const response = parts.join("\n");
    saveMessage("assistant", response, citations);
    return { response, citations, chunksUsed: context.length };
  }

  return {
    getApiKey, setApiKey, hasApiKey,
    getAllSources, deleteSource,
    processFile, processNote,
    getHistory, clearHistory,
    ask,
  };
})();
